rust/rbac-registration/src/registration/cardano/mod.rs

```diff
@@ -37,38 +40,118 @@ pub struct RegistrationChain {
+    /// `cip509.report()`.
     ///
     /// # Errors
+    ///  - Propagates any I/O or provider-level errors encountered while checking key
+    ///    ownership (e.g., database lookup failures).
+    pub async fn new<State>(
+        cip509: &Cip509,
+        state: &mut State,
+    ) -> anyhow::Result<Option<Self>>
+    where
+        State: RbacChainsState,
+    {
+        let Some(new_chain) = Self::new_stateless(cip509) else {
+            return Ok(None);
+        };
+
+        // Verify that a Catalyst ID of this chain is unique.
+        {
+            let cat_id = new_chain.catalyst_id();
+            if state.is_chain_known(cat_id).await? {
+                cip509.report().functional_validation(
+            &format!("{} is already used", cat_id.as_short_id()),
+                "It isn't allowed to use same Catalyst ID (certificate subject public key) in multiple registration chains",
+                    );
+            }
+
+            check_signing_public_key(cat_id, cip509, state).await?;
+        }
+
+        if cip509.report().is_problematic() {
+            return Ok(None);
+        }
+
+        state
+            .take_stake_address_from_chains(cip509.stake_addresses())
+            .await?;
+
+        Ok(Some(new_chain))
+    }
+
+    /// Create a new instance of registration chain.
+    /// The first new value should be the chain root.
     ///
-    /// Returns an error if data is invalid
+    /// Returns `None` if the `cip509` is invalid by any reason, properly updating
+    /// `cip509.report()`.
     #[must_use]
-    pub fn new(cip509: Cip509) -> Option<Self> {
+    pub fn new_stateless(cip509: &Cip509) -> Option<Self> {
         let inner = RegistrationChainInner::new(cip509)?;
 
         Some(Self {
             inner: Arc::new(inner),
         })
     }
 
-    /// Update the registration chain.
+    /// Attempts to update an existing RBAC registration chain
+    /// with a new CIP-509 registration, validating address and key usage consistency.
     ///
-    /// # Arguments
-    /// - `cip509` - The CIP509.
+    /// Returns `None` if the `cip509` is invalid by any reason, properly updating
+    /// `cip509.report()`.
     ///
     /// # Errors
+    ///  - Propagates any I/O or provider-level errors encountered while checking key
+    ///    ownership (e.g., database lookup failures).
+    pub async fn update<State>(
+        &self,
+        cip509: &Cip509,
+        state: &State,
+    ) -> anyhow::Result<Option<Self>>
+    where
+        State: RbacChainsState,
+    {
+        let Some(new_chain) = self.update_stateless(cip509) else {
+            return Ok(None);
+        };
+
+        // Check that addresses from the new registration aren't used in other chains.
+        let previous_addresses = self.stake_addresses();
+        let reg_addresses = cip509.stake_addresses();
+        let new_addresses: Vec<_> = reg_addresses.difference(&previous_addresses).collect();
```

Comment:
I think adding `stake` will be more readable when glancing through
eg. `previous_stake_addresses`